#### Why Data Cleaning is Important
Without properly cleaned data, the results of any data analysis or machine learning model could be inaccurate. It is a commonly known fact that data scientists spend 80% of their time cleaning, manipulating and transforming data into the shape they want in order to carry out accurate analysis.

##### Some common problems with data:
- Column headers are variables, not variable names
- Multiple variables are stored in one column
- Variables are stored in both rows and columns
- Multiple types of observational units are stored in the same table
- A single observational unit stored in multiple tables

In this project efforts will be made to deal with and overcome some common untidy data problems.

In [1]:
#Importing required libraries
import numpy as np
import pandas as pd
from random import randint
import matplotlib.pyplot as plt

In [2]:
#Loading the dataset into a dataframe called ride_sharing
ride_sharing = pd.read_csv('../Datasets/ride_sharing_new.csv')
#Viewing the first 3 rows of the dataset
ride_sharing.head(3)

Unnamed: 0    duration  station_A_id  \
0           0  12 minutes            81   
1           1  24 minutes             3   
2           2   8 minutes            67   

                                      station_A_name  station_B_id  \
0                                 Berry St at 4th St           323   
1       Powell St BART Station (Market St at 4th St)           118   
2  San Francisco Caltrain Station 2  (Townsend St...            23   

                    station_B_name  bike_id  user_type  user_birth_year  \
0               Broadway at Kearny     5480          2             1959   
1  Eureka Valley Recreation Center     5193          2             1965   
2    The Embarcadero at Steuart St     3652          3             1993   

  user_gender  
0        Male  
1        Male  
2        Male

The dataset above is a bicycle ride sharing data from San Francisco. It contains information on the start and end stations, the trip duration, and some user information for a bike sharing service.

The user_type column contains information on whether a user is taking a free ride and takes on the following values:

- 1 for free riders.

- 2 for pay per ride.

- 3 for monthly subscribers.  

In [3]:
#Printing information about the ride_sharing dataset
print(ride_sharing.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25760 entries, 0 to 25759
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Unnamed: 0       25760 non-null  int64 
 1   duration         25760 non-null  object
 2   station_A_id     25760 non-null  int64 
 3   station_A_name   25760 non-null  object
 4   station_B_id     25760 non-null  int64 
 5   station_B_name   25760 non-null  object
 6   bike_id          25760 non-null  int64 
 7   user_type        25760 non-null  int64 
 8   user_birth_year  25760 non-null  int64 
 9   user_gender      25760 non-null  object
dtypes: int64(6), object(4)
memory usage: 2.0+ MB
None


It can be observed that the dataset has 10 columns 25760 rows and mixed data types. The user_type column should be categorical and nto integer.

In [4]:
#Summary statistics of the user_type column
print(ride_sharing['user_type'].describe())

count    25760.000000
mean         2.008385
std          0.704541
min          1.000000
25%          2.000000
50%          2.000000
75%          3.000000
max          3.000000
Name: user_type, dtype: float64


The Summary Statistics shown above is not very informative.

In [6]:
#Converting the user_type column to category and storing it in a new column 'user_type_cat'
ride_sharing['user_type_cat'] = ride_sharing['user_type'].astype('category')

#Confirming the change with an assert statement. The assert statement will return nothing if the assertion is true and an error otherwise
assert ride_sharing['user_type_cat'].dtype == 'category'

print(ride_sharing['user_type_cat'].describe())

count     25760
unique        3
top           2
freq      12972
Name: user_type_cat, dtype: int64


The Summary Statistics printed now is more useful because it gives the following information:
- there 3 unique categories
- 